cli/running/running.go

```diff
@@ -124,62 +132,163 @@ func (provider *providerImpl) IsRestartable() (bool, error) {
 		return false, err
 	}
 
-	return provider.cmdCtx.Running.Restartable, nil
+	return provider.instance.Restartable, nil
+}
+
+// findInstSeparator returns instance separator index.
+// Cartridge application uses dot and dash sepatator for the application
+// and instance name (dash for stateboard, dot for others).
+func findInstSeparator(inst string) int {
+	sepIdx := -1
+	dotIdx := strings.Index(inst, ".")
+	dashIdx := strings.Index(inst, "-")
+
+	if dotIdx+dashIdx != -2 { // Separator is found.
+		mult := dotIdx * dashIdx
+		if mult < 0 { // Only one separator is found.
+			sepIdx = -mult
+		} else {
+			if dotIdx < dashIdx {
+				sepIdx = dotIdx
+			} else {
+				sepIdx = dashIdx
+			}
+		}
+	}
+
+	return sepIdx
 }
 
-// findAppFile searches of an application init file.
-func findAppFile(appName string, cliOpts *config.CliOpts) (string, error) {
+// collectInstances searches all instances available in application.
+func collectInstances(appName string, cliOpts *config.CliOpts,
+	basePath string) ([]cmdcontext.RunningCtx, error) {
 	var err error
-	appDir := cliOpts.App.InstancesAvailable
+	var appDir string
+
 	if appDir == "" {
 		if appDir, err = os.Getwd(); err != nil {
-			return "", err
+			return nil, err
 		}
+	} else if !filepath.IsAbs(appDir) {
+		appDir = filepath.Join(basePath, appDir)
 	}
 
 	var appPath string
+	singleApp := true
+
+	// The user can specify a specific instance from the application.
+	// Example: `tt status application:server`.
+	singleInst := false
+	singleInstName := ""
+	colonIds := strings.Index(appName, ":")
+	if colonIds != -1 {
+		singleInst = true
+		appNameTmp := appName
+		appName = appNameTmp[:colonIds]
+		singleInstName = appNameTmp[colonIds+1:]
+	}
 
 	// We considering several scenarios:
 	// 1) The application starts by `appName.lua`
 	// 2) The application starts by `appName/init.lua`
-	appAbsPath, err := util.JoinAbspath(appDir, appName+".lua")
-	if err != nil {
-		return "", err
-	}
-	dirAbsPath, err := util.JoinAbspath(appDir, appName)
-	if err != nil {
-		return "", err
-	}
+	// 3) The application starts by `dirName:appName`
+	// 4) Read application list from `appName/instances.yml`
+	luaPath := filepath.Join(appDir, appName+".lua")
+	dirPath := filepath.Join(appDir, appName)
 
 	// Check if one or both file and/or directory exist.
-	_, fileStatErr := os.Stat(appAbsPath)
-	dirInfo, dirStatErr := os.Stat(dirAbsPath)
+	_, fileStatErr := os.Stat(luaPath)
+	dirInfo, dirStatErr := os.Stat(dirPath)
 
 	if !os.IsNotExist(fileStatErr) {
 		if fileStatErr != nil {
-			return "", fileStatErr
+			return nil, fileStatErr
 		}
-		appPath = appAbsPath
+		appPath = luaPath
 	} else if dirStatErr == nil && dirInfo.IsDir() {
-		appPath = path.Join(dirAbsPath, "init.lua")
-		if _, err = os.Stat(appPath); err != nil {
-			return "", err
+		// Search for instances.yml
+		instCfgPath := path.Join(dirPath, "instances.yml")
+		if _, err = os.Stat(instCfgPath); err == nil {
+			// Get instances.
```

Comment:
What do you think about moving the material to get the ctx array for multiinnstace into a separate function? Now it's hard to read.

Reply:
Done